jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DoSHandler.java

```diff
@@ -255,372 +253,114 @@ protected void doStop() throws Exception
-            }
-
-            @Override
-            public boolean onRequest(long now)
-            {
-                while (true)
+                try (AutoLock ignored = _lock.lock())
                 {
-                    long state = _state.get();
-                    int encodedSampleStart = AtomicBiInteger.getHi(state);
-                    if (encodedSampleStart == 0)
-                        encodedSampleStart = encodeNanoTime(now);
-
-                    int requests = AtomicBiInteger.getLo(state);
-
-                    // Remove any drips from the bucket
-                    long startSample = decodeNanoTime(encodedSampleStart, now);
-                    long elapsed = NanoTime.elapsed(startSample, now);
-                    if (elapsed >= _nanosPerDrip)
-                    {
-                        requests = Math.max(0, requests - (int)(elapsed / _nanosPerDrip));
-                        encodedSampleStart = encodeNanoTime(now);
-                    }
-
-                    // If we have requests to spare?
-                    if (requests < _requestsPerSample)
-                    {
-                        if (_state.compareAndSet(state, encodedSampleStart, requests + 1))
-                            return true;
-                        continue;
-                    }
-
-                    // The bucket is full
-                    if (elapsed < _nanosPerDrip || _state.compareAndSet(state, encodedSampleStart, requests))
-                        return false;
+                    return _expireNanoTime;
                 }
             }
 
             @Override
-            public boolean isIdle(long now)
+            public boolean onRequest(long now)
             {
-                while (true)
+                try (AutoLock ignored = _lock.lock())
                 {
-                    long state = _state.get();
-                    int encodedSampleStart = AtomicBiInteger.getHi(state);
-                    int requests = AtomicBiInteger.getLo(state);
-
-                    long startSample = decodeNanoTime(encodedSampleStart, now);
-                    long elapsed = NanoTime.elapsed(startSample, now);
-
-                    if (requests > 0)
-                    {
-                        // We are not idle, so just remove drips
-                        if (elapsed >= _nanosPerDrip)
-                        {
-                            requests = Math.max(0, requests - (int)(elapsed / _nanosPerDrip));
-                            if (_state.compareAndSet(state, encodeNanoTime(now), requests))
-                                return false;
-                            continue;
-                        }
-                    }
-
-                    return  elapsed > _samplePeriod.toNanos();
+                    long elapsedSinceLastDrip = NanoTime.elapsed(_lastDrip, now);
+                    long drips = elapsedSinceLastDrip / _nanosPerDrip;
+                    _lastDrip = _lastDrip + drips * _nanosPerDrip;
```

Comment:
This line is `_lastDrip = now`, rather than all the math.

Reply:
That will make the algorithm numerically insensitive because of rounding errors.  If we were 1 nanosecond short of another drip, then setting now will effectively lose that drip and the rate will be slow.